### PR TITLE
RATIS-1948. NettyClientStreamRpc.Connection.scheduleReconnect should check isClosed.

### PR DESCRIPTION
**Pull request by szetszwo**
See RATIS-1948

### CODE REVIEW

**Review by adoroszlai — APPROVED** [review]
Thanks @szetszwo for updating the patch.

File: ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java [hunk 1]
```diff
@@ -277,6 +278,11 @@ public Set<ClientInvocationId> remove(ChannelId channelId) {
     this.nettyServerStreamRpcMetrics = metrics;
   }
 
+  void shutdown() {
+    requestExecutor.shutdown();
+    writeExecutor.shutdown();
```
[the remark this answers is not in the document]
Comment (szetszwo) [reply ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java]: You are right.  We should shutdownAndWait.

File: ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java [hunk 2]
```diff
@@ -260,6 +266,8 @@ synchronized boolean write(DataStreamRequest request) {
 
     synchronized boolean shouldFlush(boolean force, int countMin, SizeInBytes bytesMin) {
       if (force || count >= countMin || bytes >= bytesMin.getSize()) {
+        LOG.debug("flush: force? {}, (count, bytes)=({}, {}), min=({}, {})",
+            force, count, bytes, countMin, bytesMin);
```
Comment (adoroszlai) [inline ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java:270]: `Connection refused` may still happen in the following case: when shutting down all nodes in the group, server of one node may be shutdown before peer proxy of that server in another node.

server:

```
2023-12-02 09:01:32,887 [008c6bcb-05b6-400c-87dc-2e492505ab22-NettyServerStreamRpc-bossGroup--thread1] INFO  logging.LoggingHandler (AbstractInternalLogger.java:log(148)) - [id: 0xef0ee84e, L:/0.0.0.0:15024] CLOSE
2023-12-02 09:01:32,888 [008c6bcb-05b6-400c-87dc-2e492505ab22-NettyServerStreamRpc-bossGroup--thread1] INFO  logging.LoggingHandler (AbstractInternalLogger.java:log(148)) - [id: 0xef0ee84e, L:/0.0.0.0:15024] INACTIVE
2023-12-02 09:01:32,888 [008c6bcb-05b6-400c-87dc-2e492505ab22-NettyServerStreamRpc-bossGroup--thread1] INFO  logging.LoggingHandler (AbstractInternalLogger.java:log(148)) - [id: 0xef0ee84e, L:/0.0.0.0:15024] UNREGISTERED
```

client in peer:

```
2023-12-02 09:01:32,893 [Thread-999] DEBUG client.NettyClientStreamRpc (NettyClientStreamRpc.java:shouldFlush(269)) - flush: force? true, (count, bytes)=(0, 0), min=(0, 0)
2023-12-02 09:01:32,897 [Thread-999] WARN  util.PeerProxyMap (PeerProxyMap.java:closeProxy(170)) - c8994a5a-9b0d-4fa4-993e-22bc1af8386e-NettyServerStreamRpc: Failed to close proxy for peer 008c6bcb-05b6-400c-87dc-2e492505ab22|127.0.0.1:15023, proxy class: class org.apache.ratis.client.impl.DataStreamClientImpl
org.apache.ratis.thirdparty.io.netty.channel.AbstractChannel$AnnotatedConnectException: finishConnect(..) failed: Connection refused: /127.0.0.1:15024
```

I think the problem is that during close `shouldFlush` always returns `true` due to `force=true`.  So closing the peer proxy tries to reconnect to send the empty byte buffer.

https://github.com/apache/ratis/blob/d5019c1c7b0f65952aa88f0e81780a1b32befd89/ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java#L456-L464

Should we really force flush when there is no data to be sent?

Comment (adoroszlai) [reply ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java:270]: With https://github.com/adoroszlai/ratis/commit/7dec44ddf6739d9910cb52d638c1233e3e9977a4 (on top of this PR) `Connection refused` no longer happens.

(There are some CI failures, checking if they are related.)

Comment (adoroszlai) [reply ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java:270]: https://github.com/adoroszlai/ratis/commit/4d4760a1f41ff9721bf30951506699cdef76ee12 fixes tests: https://github.com/adoroszlai/ratis/actions/runs/7069737291

Comment (szetszwo) [reply ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java]: @adoroszlai , you are right -- it should flush only if `bytes > 0`.  Updated.

Comment (adoroszlai) [reply ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java]: @szetszwo After I fixed `shouldFlush` (slightly differently, but I think the result is the same), I got failures in `TestNettyDataStreamChainTopologyWithGrpcCluster` and `TestNettyDataStreamStarTopologyWithGrpcCluster`.  I guess the problem happens if `close()` is called without sending anything.  See my [previous comment](https://github.com/apache/ratis/pull/977#discussion_r1412783939) on how I fixed these tests.

**Comment by adoroszlai** [comment]
@szetszwo thanks for updating the patch.

Although CI says "All checks have passed", it also says `unit (server)` was skipped.  Actually it was not skipped, but could not be completed due to `No space left on device`, both in PR and fork runs:

https://github.com/apache/ratis/actions/runs/7073614362/job/19253756489?pr=977
https://github.com/szetszwo/ratis/actions/runs/7073613972/job/19253755570

If we change the new TRACE log level setting back to INFO, `TestNettyDataStreamChainTopologyWithGrpcCluster` simply times out.

File: ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java [hunk 3]
```diff
@@ -242,29 +249,44 @@ public String toString() {
...
+      if (options.contains(StandardWriteOption.CLOSE)) {
+        // flush in order to send the CLOSE option.
+        return true;
+      } else if (bytes == 0 && count == 0) {
+        // nothing to flush (when bytes == 0 && count > 0, client may have written empty packets for including options)
+        return false;
+      } else {
+        return count >= countMin
+            || bytes >= bytesMin.getSize()
+            || options.contains(StandardWriteOption.FLUSH);
```
[the remark this answers is not in the document]
Comment (szetszwo) [reply ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java:269]: Yes, it is the bug --  the header was supposed to be flushed out but the `FLUSH` option was not included.  Also, there are some other cases in the tests need to have `FLUSH`.

> ... move it to a separate `enum` which also implements `WriteOption`; ...

This is a good idea but it becomes incompatible.  Let's simply check `FLUSH` when converting to proto for now.

File: ratis-client/src/main/java/org/apache/ratis/client/RaftClientConfigKeys.java [hunk 4]
```diff
@@ -117,7 +117,7 @@ static void setFlushRequestCountMin(RaftProperties properties, int flushRequestC
     String FLUSH_REQUEST_BYTES_MIN_KEY = PREFIX + ".flush.request.bytes.min";
     SizeInBytes FLUSH_REQUEST_BYTES_MIN_DEFAULT = SizeInBytes.ONE_MB;
     static SizeInBytes flushRequestBytesMin(RaftProperties properties) {
-      return getSizeInBytes(properties::getSizeInBytes, FLUSH_REQUEST_COUNT_MIN_KEY,
+      return getSizeInBytes(properties::getSizeInBytes, FLUSH_REQUEST_BYTES_MIN_KEY,
```
[the remark this answers is not in the document]
Comment (szetszwo) [reply ratis-client/src/main/java/org/apache/ratis/client/RaftClientConfigKeys.java:120]: RATIS-1157 was to buffer the small requests but this bug made it never working.

**Comment by guohao-rosicky** [comment]
sure.
> @guohao-rosicky , please feel free to review/test this change. We could file follow up JIRAs if you find anything.

**Pull request merged**